### ISSUE TITLE
Test `test_install_legacy_yaml` is silently broken

### ISSUE THREAD

**Issue by haampie** [LLNL/spack#28779]
Ping @nhanford, can you please fix `test_install_legacy_yaml`?

It's broken yet it passes. Try this diff:

```diff
$ git diff
diff --git a/lib/spack/spack/main.py b/lib/spack/spack/main.py
index 34ba3c3349..613ba569f8 100644
--- a/lib/spack/spack/main.py
+++ b/lib/spack/spack/main.py
@@ -608,7 +608,7 @@ def __call__(self, *argv, **kwargs):
 
         out = StringIO()
         try:
-            with log_output(out):
+            with log_output(out, echo=True):
                 self.returncode = _invoke_command(
                     self.command, self.parser, args, unknown)
 
diff --git a/lib/spack/spack/test/bindist.py b/lib/spack/spack/test/bindist.py
index e45d2eb07d..b16835cade 100644
--- a/lib/spack/spack/test/bindist.py
+++ b/lib/spack/spack/test/bindist.py
@@ -619,6 +619,7 @@ def test_install_legacy_yaml(test_legacy_mirror, install_mockery_mutable_config,
                 + '/build_cache/test-debian6-core2-gcc-4.5.0-zlib-' +
                 '1.2.11-t5mczux3tfqpxwmg7egp7axy2jvyulqk.spec.yaml')
     uninstall_cmd('-y', '/t5mczux3tfqpxwmg7egp7axy2jvyulqk')
+    assert False
```

and notice that:

```console
$ spack unit-test ./lib/spack/spack/test/bindist.py -k test_install_legacy
-------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------
==> Installing zlib-1.2.11-t5mczux3tfqpxwmg7egp7axy2jvyulqk
==> Fetching file:///tmp/pytest-of-harmen/pytest-48/test_install_legacy_yaml0/legacy_yaml_mirror/build_cache/test-debian6-core2/gcc-4.5.0/zlib-1.2.11/test-debian6-core2-gcc-4.5.0-zlib-1.2.11-t5mczux3tfqpxwmg7egp7axy2jvyulqk.spack
==> Extracting zlib-1.2.11-t5mczux3tfqpxwmg7egp7axy2jvyulqk from binary cache
==> Warning: Dependency missing: may be deprecated or corrupted:
  /tmp/pytest-of-harmen/pytest-48/test_install_legacy_yaml0/opt/test-debian6-core2/gcc-4.5.0/zlib-1.2.11-t5mczux3tfqpxwmg7egp7axy2jvyulqk
  Specs zlib@1.2.11%gcc@4.5.0+optimize+pic+shared arch=test-debian6-core2 and zlib@1.2.11%apple-clang@11.0.0+optimize+pic+shared arch=test-debian6-core2 have the same SHA-1 prefix!
[+] /tmp/pytest-of-harmen/pytest-48/test_install_legacy_yaml0/opt/test-debian6-core2/gcc-4.5.0/zlib-1.2.11-t5mczux3tfqpxwmg7egp7axy2jvyulqk
==> Successfully uninstalled zlib@1.2.11%gcc@4.5.0+optimize+pic+shared arch=test-debian6-core2/t5mczux
------------------------------------------------------------------ Captured stdout teardown ------------------------------------------------------------------
```

Notice:

```
  Specs zlib@1.2.11%gcc@4.5.0+optimize+pic+shared arch=test-debian6-core2 and zlib@1.2.11%apple-clang@11.0.0+optimize+pic+shared arch=test-debian6-core2 have the same SHA-1 prefix!
```

The issue is that the blob `test-debian6-core2-gcc-4.5.0-zlib-1.2.11-t5mczux3tfqpxwmg7egp7axy2jvyulqk.spack` lists `%apple-clang` as a compiler, but the `spec.yaml` specifies `%gcc`. So what is requested is not what is installed, and the the database / filesystem become inconsistent.

Came up in #28764

**Comment by haampie** [comment]
also pinging @scheibelp regarding #27659 cause this is another reason I'd want `echo=True` in SpackCommand

**Comment by alalazo** [comment]
Fixed in #34347
